### PR TITLE
Feat: Add support for deploying on iOS (Preview)

### PR DESCRIPTION
**Pull request by shivamj-0303**
# Feat: Add CI/CD Support for TestFlight Deployment on PRs (iOS)

## Description
This PR adds full CI automation for iOS builds deployed to TestFlight on every pull request update.  
It leverages Fastlane with Match for manual code signing, enforces unique and valid build numbers, and integrates with GitHub Actions to automate the entire process from code to TestFlight.
A cleanup workflow has been implemented to automatically expire builds associated with closed or merged pull requests.

## Why is this change needed?
- To allow QA to test iOS builds for every PR without requiring manual uploads.
- To enforce consistency in code signing, provisioning, and build versioning.
- To support better collaboration by posting TestFlight and BrowserStack links directly on PRs.
- To automate stale build cleanup and reduce clutter on TestFlight.

## What does this PR do?
- Adds `ci_signing_setup` Fastlane lane to configure signing identities and provisioning profiles using Fastlane Match.
- Adds `pr_deploy` Fastlane lane to build the iOS app and upload it to TestFlight using a structured build number format (`MMDD.HHMM.PR_NUMBER`).
- Adds GitHub Actions workflow that:
  - Runs on PR open/sync events
  - Sets up Fastlane, Node, and CocoaPods
  - Builds the app and uploads it to TestFlight
  - Posts a comment on the PR with TestFlight links
- Adds a cleanup GitHub Actions workflow triggered on PR close/merge that:
  - Authenticates with App Store Connect
  - Searches and expires builds whose build number ends with the PR number

## How was this tested?
- Opened and updated a PR → Verified successful TestFlight build upload and code signing.
- Verified build number format in App Store Connect TestFlight.
- Confirmed comment with TestFlight links was posted on the PR.
- Closed the PR → Confirmed cleanup workflow expired the correct build from TestFlight.
- Verified the app works correctly on BrowserStack.

## Screenshots
i2
i1
Screenshot from 2025-06-24 17-08-27

### CODE REVIEW

File: .github/CI_README.md
Comment (jjalan) [inline .github/CI_README.md:1]: Can you make sure we prefix with IOS_, and rename variables so they are readable?

File: ios/fastlane/Appfile
Comment (jjalan) [inline ios/fastlane/Appfile:1]: Sounds like the values should come from config

File: ios/fastlane/Matchfile
Comment (jjalan) [inline ios/fastlane/Matchfile:1]: Maybe they should come from env variables

File: .github/CI_README.md
Comment (shivamj-0303) [reply .github/CI_README.md:1]: Renamed all IOS related variables to be prefixed with IOS_ . Let me know if any additional context or adjustments are needed.

File: ios/fastlane/Appfile
Comment (shivamj-0303) [reply ios/fastlane/Appfile:1]: Moved the hardcoded values from Appfile to use ENV values instead.

File: ios/fastlane/Matchfile
Comment (shivamj-0303) [reply ios/fastlane/Matchfile:1]: Updated the Matchfile to read values from environment variables instead of hardcoding them.

File: .github/workflows/preview_on_pr_update.yml [hunk 1]
```diff
@@ -200,3 +200,143 @@ jobs:
...
+      - name: Post PR comment with TestFlight link
+        if: ${{ github.event.pull_request }}
+        env:
+          GH_TOKEN: ${{ secrets.GITHUB_TOKEN }}
+          PR_NUMBER: ${{ github.event.pull_request.number }}
+          BUILD_HASH: ${{ github.sha }}
+        run: |
+          brew install gh # Only needed if gh CLI is not pre-installed
+          gh auth setup-git
+          TESTFLIGHT_LINK="https://appstoreconnect.apple.com/teams/$APP_STORE_CONNECT_API_KEY_ISSUER_ID/apps/$APPLE_ID/testflight/ios"
```
Comment (Patel-world) [inline .github/workflows/preview_on_pr_update.yml]: You're using two env variables:
$APP_STORE_CONNECT_API_KEY_ISSUER_ID

$APPLE_ID

But they are not defined in this env: block, and hence they will be empty or undefined at runtime.

File: ios/fastlane/scripts/ios_testflight_cleanup.rb [hunk 2]
```diff
@@ -0,0 +1,40 @@
...
+      FastlaneCore::UI.message("🧹 Expiring build #{build_version_str} (#{build.id}) for PR ##{pr_number}")
+        begin
+          build = Spaceship::ConnectAPI::Build.get(build_id: build.id)
+
+        unless build
+          FastlaneCore::UI.user_error!("Build with ID #{build.id} not found.")
+        end
+          build.expire!
+        FastlaneCore::UI.success("Successfully expired build ID: #{build.id}")
+      rescue StandardError => e
```
Comment (Patel-world) [inline ios/fastlane/scripts/ios_testflight_cleanup.rb]: NIT: Indentation of begin ... rescue is a bit off.
Can you check

File: ios/fastlane/scripts/ios_testflight_deploy.rb [hunk 3]
```diff
@@ -0,0 +1,123 @@
...
+  # Strip bitcode manually from Hermes binary to avoid App Store submission errors.
+  # Hermes framework often includes bitcode sections that aren't removed by default tools.
+  sh <<~BASH
+      echo "🔍 Stripping bitcode from Hermes binary before uploading to TestFlight..."
+
+      IPA_PATH=#{lane_context[:IPA_OUTPUT_PATH]}
+      unzip -q "$IPA_PATH" -d temp_payload
+
+      HERMES_BIN="temp_payload/Payload/Boilerplate.app/Frameworks/hermes.framework/hermes"
+      APP_PATH="temp_payload/Payload/Boilerplate.app"
```
Comment (Patel-world) [inline ios/fastlane/scripts/ios_testflight_deploy.rb]: looks like hardcoded Boilerplate.app

File: ios/fastlane/scripts/ios_testflight_deploy.rb [hunk 4]
```diff
@@ -0,0 +1,123 @@
...
+    type: "appstore",
+    readonly: true,
+    verbose: true,
+    keychain_name: keychain_name,
+    keychain_password: keychain_password
+  )
+  # Set the build number using current datetime + PR number to ensure uniqueness across PR builds.
+  increment_build_number(
+    xcodeproj: xcodeproj,
+    build_number: "#{Time.now.strftime('%m%d.%H%M')}.#{ENV['PR_NUMBER']}"
```
Comment (Patel-world) [inline ios/fastlane/scripts/ios_testflight_deploy.rb]: PR_NUMBER is coming from env ??

Comment (shivamj-0303) [reply ios/fastlane/scripts/ios_testflight_deploy.rb]: Yes, PR_NUMBER is injected from the GitHub Actions workflow via env: PR_NUMBER: ${{ github.event.pull_request.number }}.

Comment (Patel-world) [reply ios/fastlane/scripts/ios_testflight_deploy.rb]: pr_number = options.fetch(:pr_number)

is that different from this one

Comment (shivamj-0303) [reply ios/fastlane/scripts/ios_testflight_deploy.rb]: Yes, I initially used ENV['PR_NUMBER'] since it’s injected from GitHub Actions as an environment variable.

However, I’ve now updated the script to use pr_number = options.fetch(:pr_number) instead

File: ios/fastlane/scripts/ios_testflight_deploy.rb (hunk 3 above)
Comment (shivamj-0303) [reply ios/fastlane/scripts/ios_testflight_deploy.rb]: updated to use dynamic app name instead of hardcoded Boilerplate.app

File: ios/fastlane/scripts/ios_testflight_cleanup.rb (hunk 2 above)
Comment (shivamj-0303) [reply ios/fastlane/scripts/ios_testflight_cleanup.rb]: Fixed the indentation around the begin ... rescue block

File: .github/workflows/preview_on_pr_update.yml (hunk 1 above)
Comment (shivamj-0303) [reply .github/workflows/preview_on_pr_update.yml]: Added missing env variables by prefixing them with IOS_ (i.e., IOS_APP_STORE_CONNECT_API_KEY_ISSUER_ID and IOS_APPLE_ID) to match the secrets defined in the workflow.

**Comment by AkshatBari02** [comment]
Is this deployment over testflight for IOS available for all PRs now?
Can we build this IOS build using Windows or Ubuntu also? If yes, then how?

**Comment by shivamj-0303** [comment]
> Is this deployment over testflight for IOS available for all PRs now?
> Can we build this IOS build using Windows or Ubuntu also? If yes, then how?

 TestFlight deployment will be available after this is merged — it's currently set up on my feature branch.
 iOS builds can't run natively on Windows or Ubuntu due to Xcode requirements.
Our GitHub Actions workflow uses macOS runners to build and upload to TestFlight, so no local Mac is needed.

File: ios/fastlane/scripts/ios_testflight_deploy.rb [hunk 5]
```diff
@@ -0,0 +1,121 @@
...
+    verbose: true,
+    keychain_name: keychain_name,
+    keychain_password: keychain_password
+  )
+  # Set the build number using current datetime + PR number to ensure uniqueness across PR builds.
+  increment_build_number(
+    xcodeproj: xcodeproj,
+    build_number: "#{Time.now.strftime('%m%d.%H%M')}.#{pr_number}"
+  )
+
```
Comment (AkshatBari02) [inline ios/fastlane/scripts/ios_testflight_deploy.rb:33]: Can we use Year (YYYY) also for making build number unique as the current build number may get repeat or collide after an year ?

Comment (shivamj-0303) [reply ios/fastlane/scripts/ios_testflight_deploy.rb:33]: Thanks Akshat! Since the pr_number is included in the build number and PRs usually don’t stay open for a year, the chance of collision is minimal. But happy to switch to '%Y%m%d.%H%M' if we want to make it more robust for edge cases. Let me know your thoughts.

Comment (AkshatBari02) [reply ios/fastlane/scripts/ios_testflight_deploy.rb:33]: Yes, PRs don't stay open for a year, but it can be a edge case and if this workflow is used in another big project or repo there also it can be useful.

Comment (shivamj-0303) [reply ios/fastlane/scripts/ios_testflight_deploy.rb:33]: That makes sense — adding the year would make it more robust and safer for reuse in other projects or edge cases. I’ll update the build number format to include the year (%Y%m%d.%H%M) to avoid any potential collisions.

**Comment by shivamj-0303** [comment]
## ✅ iOS TestFlight CI Integration: Test Summary for Recco App

Steps taken to test the new iOS TestFlight CI pipeline integration on the Recco iOS application.

---

### 🧪 Test Steps

1. **Configured Node Version for CI**
   - Updated workflows to use `.nvmrc` for consistent Node.js version across CI.
   - Set Node.js version to `22.13.1` in `.nvmrc`.

2. **Resolved C++ Compatibility in iOS Build**
   - Updated `Podfile` to explicitly set C++ standard to prevent build errors on Xcode.
   - Ensured compatibility with native dependencies requiring modern C++.

3. **Fixed CocoaPods Dependency Issues**
   - Resolved duplicate or incompatible pods causing iOS build failures.
   - Cleaned up pod cache and ran `pod install` as part of the CI.

4. **Patched Private Apple API Usage**
   - Added a `patch-package` patch to rename private selectors (e.g., `_isKeyDown`, `_modifiedInput`) in `react-native` sources.
   - Prevents App Store rejection for ITMS-90338: Non-public API usage.
   - Patch applied automatically post-install during CI.

5. **Verified TestFlight Build Upload**
   - ✅ GitHub Actions triggered and built iOS app via Fastlane.
   - ✅ Archive successfully uploaded to TestFlight.
   - ✅ Verified app was available via TestFlight and tested on BrowserStack.

6. **Tested Cleanup on PR Close**
   - ✅ On PR close, the cleanup workflow ran successfully.
   - ✅ Confirmed the corresponding TestFlight build was expired and removed.

---

### 📸 Screenshots
reco
recco1
recco3
reccoo5
recco4

---

### 📋 Notes

- Patched `node_modules` are safe and reproducible via committed patch files.
- The workflow supports both upload and cleanup lifecycle for preview TestFlight builds.
- Verified the entire E2E flow including build, upload and expiration.

---

✅ **Test Result**: **Success**  
All flows including build, upload to TestFlight, and cleanup were verified end-to-end on the Recco iOS app.

File: ios/Gemfile [hunk 6]
```diff
@@ -0,0 +1,5 @@
+source "https://rubygems.org"
+
+gem "fastlane"
+gem "cocoapods"
```
Comment (PunithKosana) [inline ios/Gemfile:4]: Where is the `Gemfile.lock` file for these dependencies?

File: .github/workflows/preview_on_pr_update.yml [hunk 7]
```diff
@@ -200,3 +200,144 @@ jobs:
...
             });
+
+  testflight_deploy:
+    name: TestFlight Preview Deploy
+    needs: [checks]
+    if: github.event.action != 'closed'
+    runs-on: macos-14 
+    timeout-minutes: 30
+    env:
+      NO_FLIPPER: 1
```
Comment (PunithKosana) [inline .github/workflows/preview_on_pr_update.yml]: What does `NO_FLIPPER: 1` refers to? Add an comment to it explaining what it does or please do mention it in the `CI_README.md`

Comment (shivamj-0303) [reply .github/workflows/preview_on_pr_update.yml]: Thanks for pointing this out! I’ve added the above inline comment to the workflow that NO_FLIPPER=1 disables all Flipper dependency installation during the build.

File: ios/Gemfile (hunk 6 above)
Comment (shivamj-0303) [reply ios/Gemfile:4]: Added Gemfile.lock in ios/Gemfile.lock

File: .github/CI_README.md [hunk 8]
```diff
@@ -32,14 +32,26 @@ These environment variables are used by the GitHub Actions workflows and Fastlan
...
+| `IOS_APP_STORE_CONNECT_API_KEY_ISSUER_ID`| GitHub Secret    | App Store Connect API Issuer ID. Used in conjunction with API Key ID and Base64 key to authenticate.                                                 |
+| `IOS_APP_STORE_CONNECT_API_KEY_B64`      | GitHub Secret    | Base64-encoded contents of your App Store Connect API key (.p8 file). Used by Fastlane for secure authentication with Apple APIs.                    |
+| `IOS_MATCH_PASSWORD`                     | GitHub Secret    | Password for the Match encryption repo (used by Fastlane Match to decrypt signing certificates and provisioning profiles).                           |
+| `IOS_MATCH_DEPLOY_KEY`                   | GitHub Secret    | SSH private key with read access to your Match certificate repository. Used to fetch provisioning profiles during CI.                                |
+| `IOS_APP_IDENTIFIER`                | GitHub Secret    | iOS app bundle identifier. Used during iOS build and upload processes.
+| `IOS_MATCH_PASSWORD`         | GitHub Secret  | The password used to encrypt/decrypt certificates and profiles in the Match repository. |
```
Comment (jjalan) [inline .github/CI_README.md]: I see this variable repeated in README.

Can you please make sure each variables are listed once, ordered alphabetically?

File: .github/workflows/preview_on_pr_update.yml [hunk 9]
```diff
@@ -200,3 +200,138 @@ jobs:
 
               🔗 [View Release in Firebase Console](${releaseUrl})`
             });
+
+  testflight_deploy:
+    name: TestFlight Preview Deploy
+    needs: [checks]
+    if: github.event.action != 'closed'
```
Comment (jjalan) [inline .github/workflows/preview_on_pr_update.yml:207]: In clean up, you also have `github.event.pull_request.merged == true`, should negative of that check be added here as well?

File: ios/.env.preview [hunk 10]
```diff
@@ -0,0 +1,6 @@
+ENVIRONMENT=preview
+API_BASE_URL=https://boilerplate-mern.platform.jalantechnologies.com/api
+
+DD_APPLICATION_ID=6755ccc7-0bf2-443e-bd4e-8df82d66f19d
```
Comment (jjalan) [inline ios/.env.preview:4]: For each client, I would assume this would change. Should this come from env?

Also, what are these variables and what do they do?

File: ios/Boilerplate.xcodeproj/project.pbxproj [hunk 11]
```diff
@@ -723,4 +723,4 @@
 /* End XCConfigurationList section */
 	};
 	rootObject = 83CBB9F71A601CBA00E9B192 /* Project object */;
-}
+}
```
Comment (jjalan) [inline ios/Boilerplate.xcodeproj/project.pbxproj]: Revert change.

File: react-native.config.js [hunk 12]
```diff
@@ -5,5 +5,8 @@ module.exports = {
         ios: null, // Disable autolinking for iOS to manually manage font files and avoid duplication issues.
       },
     },
+    ...(process.env.NO_FLIPPER === '1'
```
Comment (jjalan) [inline react-native.config.js:9]: Its not clear what this is doing, and why this change is needed.

**Review by jjalan — COMMENTED** [review]
Left some cursory comments

File: .github/CI_README.md (hunk 8 above)
Comment (shivamj-0303) [reply .github/CI_README.md]: Thanks for catching that! I will removed the duplicate IOS_MATCH_PASSWORD entry and reordered the list alphabetically so each variable now appears only once.

File: .github/workflows/preview_on_pr_update.yml (hunk 9 above)
Comment (shivamj-0303) [reply .github/workflows/preview_on_pr_update.yml:207]: Hey @jjalan, thanks for the catch!

In the cleanup workflow we’re listening only for pull_request.types: [closed], so we need github.event.pull_request.merged == true (and could add a merged == false branch) to distinguish merged vs. abandoned closes.

In the deploy workflow, by contrast, we gate the job with

if: github.event.action != 'closed'
which means “run on every PR event except when it’s closed (merged or not).” That single condition already skips both merged and un-merged closes, so there’s no need to add a merged == false check here.

We could swap it to if: github.event.pull_request.state == 'open' if you think that reads more clearly, but functionally it’s equivalent. Let me know if you’d like that tweak!

File: ios/.env.preview (hunk 10 above)
Comment (shivamj-0303) [reply ios/.env.preview:4]: These values are pulled straight from our  .env file and used at build time.

ENVIRONMENT tells the app it’s running in “preview” mode (so we can toggle feature flags, logging, etc.).

API_BASE_URL points to the preview backend endpoint.

DD_APPLICATION_ID is the Datadog client-side application ID we use for real-user monitoring and tracing.

DD_CLIENT_TOKEN is your public RUM/logs ingestion token, used by the iOS SDK to authenticate requests.

DD_SITE tells the SDK which Datadog intake to use (e.g. US1, US5, EU1, etc.).

whenever we onboard a new client or need a different staging setup, we’d simply add a new .env.preview (or override these values in CI), so no hardcoding is necessary here. Let me know if you’d like any changes.

File: react-native.config.js (hunk 12 above)
Comment (shivamj-0303) [reply react-native.config.js:9]: Thanks for pointing this out! We’re disabling Flipper autolinking here so that our Podfile’s NO_FLIPPER flag (which toggles Flipper via FlipperConfiguration.disabled/enabled) is the single source of truth for including or excluding Flipper dependencies. This avoids having them linked twice or unexpectedly when NO_FLIPPER=1 is set.
(Added comment for the same)

File: ios/Boilerplate.xcodeproj/project.pbxproj (hunk 11 above)
Comment (shivamj-0303) [reply ios/Boilerplate.xcodeproj/project.pbxproj]: Done

File: .github/workflows/preview_on_pr_update.yml (hunk 9 above)
Comment (jjalan) [reply .github/workflows/preview_on_pr_update.yml:207]: As you see fit.

File: ios/fastlane/scripts/ios_testflight_deploy.rb (hunk 5 above)
Comment (AkshatBari02) [reply ios/fastlane/scripts/ios_testflight_deploy.rb:33]: Please update the description of PR too.

File: .github/workflows/preview_on_pr_update.yml [hunk 13]
```diff
@@ -200,3 +200,138 @@ jobs:
...
+        working-directory: ios
+        run: |
+          rm -f Podfile.lock
+          bundle exec pod install --repo-update
+
+      # Set up SSH key so Match can pull signing identities from private Git repo
+      - name: Set up SSH for Match
+        run: |
+          mkdir -p ~/.ssh
+          echo "${{ secrets.IOS_MATCH_DEPLOY_KEY }}" > ~/.ssh/id_rsa
```
Comment (amaan-sayyed) [inline .github/workflows/preview_on_pr_update.yml]: Optional: We could simplify the SSH setup by using webfactory/ssh-agent@v0.7.0, which avoids manual key file handling and makes the setup more portable and secure.

File: .github/workflows/preview_on_pr_update.yml [hunk 14]
```diff
@@ -200,3 +200,138 @@ jobs:
...
+          skip && /\}/ { skip=0; next }
+          skip && /:sha256/ { next }
+          { print }
+          ' boost.podspec > boost.podspec.tmp && mv boost.podspec.tmp boost.podspec
+
+      - name: Install CocoaPods dependencies
+        working-directory: ios
+        run: |
+          rm -f Podfile.lock
+          bundle exec pod install --repo-update
```
Comment (amaan-sayyed) [inline .github/workflows/preview_on_pr_update.yml:272]: Curious if there's a specific reason we're deleting Podfile.lock before installing? Keeping it may help ensure consistent dependency resolution, especially for CI reproducibility.

File: ios/fastlane/scripts/ios_testflight_deploy.rb [hunk 15]
```diff
@@ -0,0 +1,122 @@
...
+      rm -rf temp_payload
+    BASH
+  # Upload the build to TestFlight (internal only) with a changelog indicating the PR number.
+  upload_to_testflight(
+    changelog: "PR ##{pr_number} Build - automated upload",
+    distribute_external: false,
+    username: username,
+    apple_id: apple_id,
+    app_identifier: app_identifier
+  )
```
Comment (amaan-sayyed) [inline ios/fastlane/scripts/ios_testflight_deploy.rb]: There’s no error handling if upload_to_testflight fails. This could silently break preview deployments and leave expired .ipas around. Add begin-rescue blocks or use fastlane's error/after_all hooks to ensure partial failures don’t leave the state inconsistent. For example:
Cleanup any temp IPA directories
Alert the user that upload failed
Optionally trigger ios_testflight_cleanup to remove stale builds

File: .github/workflows/preview_on_pr_update.yml (hunk 14 above)
Comment (shivamj-0303) [reply .github/workflows/preview_on_pr_update.yml:272]: Remove the Podfile.lock to ensure that CocoaPods resolves the correct versions of dependencies,
particularly the DatadogTrace pod, which has a version conflict between the lock file and the required version
by DatadogSDKReactNative (~> 2.27.0). This allows CocoaPods to install the necessary versions and regenerate
the Podfile.lock with the compatible versions.

File: .github/workflows/preview_on_pr_update.yml (hunk 13 above)
Comment (shivamj-0303) [reply .github/workflows/preview_on_pr_update.yml]: I’ve updated the workflow to use webfactory/ssh-agent@v0.7.0 for setting up SSH.

File: ios/fastlane/scripts/ios_testflight_deploy.rb (hunk 15 above)
Comment (shivamj-0303) [reply ios/fastlane/scripts/ios_testflight_deploy.rb]: I’ve added error handling around the upload_to_testflight step to ensure that failures are caught and handled properly. In case of failure, any temporary IPA files will be cleaned up, and the user will be alerted. Additionally, I’ve included the ios_testflight_cleanup to remove any stale builds if necessary.

**Review by PunithKosana — APPROVED** [review]

**Review by amaan-sayyed — APPROVED** [review]

File: .github/workflows/preview_on_pr_update.yml [hunk 16]
```diff
@@ -200,3 +200,143 @@ jobs:
...
+
+      - name: Install Bundler
+        run: gem install bundler
+      
+      - name: Install Ruby dependencies
+        working-directory: ios
+        run: bundle install
+      
+      - name: Set Node Version
+        uses: actions/setup-node@v4
```
Comment (Patel-world) [inline .github/workflows/preview_on_pr_update.yml:244]: Use actions/setup-node@v4 consistently
I've seen setup-node@v3 in this file, Prefer v4 for consistency and latest LTS behavior.

File: .github/workflows/preview_on_pr_update.yml [hunk 17]
```diff
@@ -200,3 +200,143 @@ jobs:
...
+
+      # Comment on the PR with a link to the TestFlight build in App Store Connect
+      - name: Post PR comment with TestFlight link
+        if: ${{ github.event.pull_request }}
+        env:
+          GH_TOKEN: ${{ secrets.GITHUB_TOKEN }}
+          PR_NUMBER: ${{ github.event.pull_request.number }}
+          BUILD_HASH: ${{ github.sha }}
+        run: |
+          brew install gh # Only needed if gh CLI is not pre-installed
```
Comment (Patel-world) [inline .github/workflows/preview_on_pr_update.yml:333]: brew install gh in CI can be slow. GitHub Actions runner already comes with gh pre-installed. Only install if it's not available:

`command -v gh || brew install gh`
